report/3.Theory/TypeRules.tex

```diff
@@ -1,13 +1,373 @@
+
+\begin{center}
+	\begin{math}
+	\cfrac
+	{E \vdash e_1 : T\ .
+		.
+		.
+		\ E \vdash e_k : T}
+	{E \vdash [e_1,\ .
+		.
+		.
+		\ , \ e_k] : list[T]}
+	\end{math}
+	\\[1\baselineskip]
+	\texttt{where} $T \notin \{none\}$
+	\texttt{and} $k > 0$
+\end{center}
+
+\textbf{Tuple literal type}\\
+The tuple type allows each entry to have any type that isn't $none$ or $generic$.
+
+\begin{center}
+	\begin{math}
+	\cfrac
+	{E \vdash e_1 : T_1,\ .
+	.
+	.
+	\ , \ E \vdash e_k : T_k}
+	{E \vdash (e_1,\ .
+		.
+		.
+		\ ,\ e_k) : tuple[T_1, \ .
+		.
+		.
+		\ ,\ T_k]}
+	\end{math}
+	\\[1\baselineskip]
+	\texttt{where} $T_i \notin \{none\}$\\
+	\texttt{and} $k > 0$\\
+	\texttt{and} $1 \le i \le k$
+\end{center}
+
+\textbf{Function type}\\
+The function type allows parameters of any type that isn't $none$, and a return type of any type.
+
+\begin{center}
+	\begin{math}
+	E \vdash f : func[T_1,\ .
+	.
+	.
+	\ ,\ T_k]
+	\end{math}
+	\\[1\baselineskip]
+	\texttt{where for each} $T \in \{T_1,\ .
+	.
+	.
+	\ ,\ T_{k-1}\},\ T \notin \{none\}$\\
+	\texttt{and} $k > 0$
+\end{center}
+
+\textbf{In-line lambda expression type}\\
+The inline lambda expression is special in that the return type is inferred from the return statement in the expression, but the types of the parameters in the expression are determined by the function whom the lambda expression is passed to.
+
+\begin{center}
+	\begin{math}
+	E \vdash \{\ S\ \} : func[T_1,\ .
+	.
+	.
+	\ ,\ T_k]
+	\end{math}
+	\\[1\baselineskip]
+	\texttt{where} $T_k$ is the type of the return expression\\
+	\texttt{and} $T_1, \ .
+	.
+	.
+	\ ,\ T_{k-1}$ are inferred from the function whom the lambda expression is passed to%Jeg er ikke sikker på, om denne sætning beskriver det godt nok
+\end{center}
+
+\textbf{Function call}\\
+A function call evaluates to a value of its return type, if the arguments for the call match the types specified in the function's declaration.
+
+\begin{center}
+	\begin{math}
+		\cfrac
+		{E \vdash e_1 : T_1 \ .
+		.
+		.
+		\ E \vdash e_k : T_k}
+		{E \vdash e_1\ f\ e_2 \ .
+		.
+		.
+		\ e_k : T_f}
+	\end{math}
+	\\[1\baselineskip]
+	\texttt{where} $f$ is a function identifier\\
+	$e_1\ .
+	.
+	.
+	\ e_k$ are the expressions passed to function $f$\\
+	\texttt{and} $T_1\ .
+	.
+	.
+	\ T_k$ are the types of the formal parameters of $f$\\
+	\texttt{and} $T_f$ is the return type of $f$\\
+	\texttt{and} $k \ge 0$
+\end{center}
+
+
 
 \subsection{Type Rules for Statements}
 %quick introduction.
+The type rules for statements are, as mentioned earlier, different from the rules for expressions.
+Statements are not evaluated to a type.
+Rather, they only determine if a program is type valid.
+If a statement is type valid, it is evaluated to \texttt{ok}.
+Otherwise, the compiler with throw an error.
+
+Just like the rules for expressions, the type rules for statements have been simplified to not include any function calls.
```

Comment:
hvorfor er de simplified though?

Reply:
Måske er "simplified" det forkerte ord. Det betyder bare, at vi ikke har specifikke typeregler for fx kald af then, while, etc., fordi deres regler allerede er dækket af reglen for funktionskald